crates/motoko/src/lib/vm.rs

```diff
@@ -671,38 +674,39 @@ fn exp_step(core: &mut Core, exp: Exp_) -> Result<Step, Interruption> {
     }
 }
 
-fn pattern_matches(env: &Env, pat: &Pat, v: &Value) -> Option<Env> {
-    match (pat, v) {
-        (Pat::Wild, _) => Some(env.clone()),
-        (Pat::Literal(Literal::Unit), Value::Unit) => Some(env.clone()),
-        (Pat::Paren(p), v) => pattern_matches(env, &*p.0, v),
-        (Pat::Annot(p, _), v) => pattern_matches(env, &*p.0, v),
-        (Pat::Var(x), v) => {
-            let mut env = env.clone();
-            env.insert(*x.0.clone(), v.clone());
+fn pattern_matches(env: Env, pat: &Pat, v: Value_) -> Option<Env> {
```

Comment:
Should `pattern_matches` really take ownership over `v : Value_`, or take the value as `v : &Value_` instead?

I'm unsure myself -- this design space feels a little big.

But my intuition is that we want to pass things by reference, knowing that we can take ownership over a shared version of them anytime need to using `fast_clone`/`share`.

WDYT?

Reply:
Haha, wow, I'm glad you noticed this. I've been going back and forth on this all day. I was originally passing by reference, but they each have slight tradeoffs in terms of the final amount of overall clones. 

We can benchmark this at some point (arbitrarily going with this for now since it appears slightly more efficient for the `canister-profiling` use case).

Reply:
Okay, well if you are actually measuring things, please, by all means, be guided by that.  As I said, it seems very unclear to me which is better, so I'd be inclined to defer to what the numbers say as well.

I should mention how much these choices have been in my head over the years. (trying my best to do FP-style programming in Rust since day 1, basically, and fighting Rust ever since).  It's a little disappointing that I haven't gotten the choices all clear in my mind yet. 

OTOH, finally doing this VM project with you, with these principles in mind, feels healing and cathartic at many points as well, for the same reasons.

It's comforting to know that you also find the choices unclear!